### PR TITLE
Fix coverity 1506297 & 1506298: negative returns

### PR DESCRIPTION
**Pull request by paulidale**
Two more that had previously been ignored.

### CODE REVIEW

File: crypto/dh/dh_ameth.c
Comment (t8m) [inline crypto/dh/dh_ameth.c:210]: why not <= 0? IMO 0 bytes private key does not make any sense.

File: crypto/dsa/dsa_ameth.c
Comment (t8m) [inline crypto/dsa/dsa_ameth.c:201]: Same as above.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Ping @openssl/committers for second review

**Comment by slontis** [comment]
In dh_pub_encode() it frees the ptr also - not sure if there is a reason for this..

**Comment by paulidale** [comment]
> In dh_pub_encode() it frees the ptr also - not sure if there is a reason for this.

Huh?  This isn't about that function.

**Comment by slontis** [comment]
> > In dh_pub_encode() it frees the ptr also - not sure if there is a reason for this.
> 
> Huh? This isn't about that function.

dh_pub_encode != dh_priv_encode (Just saying they have different patterns in the err: blocks)

**Review by slontis — APPROVED** [review]

**Pull request closed without merging**
